Test book controller

Drocheg opened the pull request:
-Adds user factory
-Adds pry and awesome_print gems 
-Adds a factory for rents
-Adds multiple tests for the book controller index and show routes

https://trello.com/c/rvPAQK1G/25-crear-tests-para-rutas-index-y-show-de-book

Drocheg committed:
Adds user factory, pry and awesome_print gems and book:controller test

Drocheg committed:
Adds rent factory and book-controller testing

-Adds a factory for rents
-Adds multiple tests for the book controller index and show routes

Gmarn commented on spec/controllers/authenticated_user.rb:
```diff
@@ -0,0 +1,12 @@
+
+RSpec.shared_context 'Authenticated User' do
+  let(:user) { create(:user) }
+  let!(:access_data) { AuthenticableEntity.generate_access_token(user) }
```
You do not need the `!` here.

Gmarn commented on spec/controllers/authenticated_user.rb:
```diff
@@ -0,0 +1,12 @@
+RSpec.shared_context 'Authenticated User' do
+  let(:user) { create(:user) }
+  let!(:access_data) { AuthenticableEntity.generate_access_token(user) }
+  let!(:access_token) { access_data[:token] }
```
Same

Gmarn commented on spec/controllers/books_controller_spec.rb:
```diff
@@ -0,0 +1,112 @@
+    end
+
+    context 'When fetching a book that exist with a rent that is not current' do
+      let(:book) do
```
You can write this on one line:
`let(:book) { create(:book) }`
Unless rubocop do no allow you this

Gmarn commented on spec/controllers/books_controller_spec.rb:
```diff
@@ -0,0 +1,112 @@
+      let(:book) do
+        create(:book)
+      end
+      let(:rent) do
```
same

Gmarn commented on spec/controllers/books_controller_spec.rb:
```diff
@@ -0,0 +1,112 @@
+    end
+
+    context 'When fetching a book that exist with a rent that is current' do
+      let(:book) do
```
same

Gmarn commented on spec/controllers/books_controller_spec.rb:
```diff
@@ -0,0 +1,112 @@
+        create(:book)
+      end
+      let!(:rent) do
+        create(:rent, book: book, user: user, start_date: Time.zone.now - 10.months \
```
preferable to not use this syntaxis to cut line, better something like that:
```ruby
create(:rent, book: book, 
              user: user, 
              start_date: Time.zone.now - 10.months,
              end_date: Time.zone.now + 10.months)
```

Gmarn commented on spec/factories.rb:
```diff
@@ -4,9 +4,24 @@
     year { Faker::Number.number(4) }
   end
+
+  factory :user do
```
It not a good pratice to put all factories in the same file. Better create a repository `factories/`, and in this one have a file by model. `users.rb` `books.rb` and `rents.rb`

Gmarn replied:
You did not separate factories in multiple files.

Gmarn commented on spec/factories.rb:
```diff
@@ -4,9 +4,24 @@
+  factory :user do
+    first_name { Faker::Name.first_name }
+    last_name { Faker::Name.last_name }
+    password { 123_123 }
```
You can use faker to generate the password.

Gmarn commented on spec/factories.rb:
```diff
@@ -4,9 +4,24 @@
+    first_name { Faker::Name.first_name }
+    last_name { Faker::Name.last_name }
+    password { 123_123 }
+    password_confirmation { 123_123 }
```
And here call: `password_confirmation { password }`

Gmarn commented on spec/factories.rb:
```diff
@@ -4,9 +4,24 @@
+  end
+
+  factory :rent do
+    user_id { Faker::Number.number(1000) }
```
Not a good way, better like that: `user { create(:user) }`

Gmarn commented on spec/factories.rb:
```diff
@@ -4,9 +4,24 @@
+
+  factory :rent do
+    user_id { Faker::Number.number(1000) }
+    book_id { Faker::Number.number(1000) }
```
Same with book

Gmarn commented on spec/factories.rb:
```diff
@@ -4,9 +4,24 @@
+  factory :rent do
+    user_id { Faker::Number.number(1000) }
+    book_id { Faker::Number.number(1000) }
+    start_date { Faker::Time.between(Time.now.utc - 1000.months, Time.now.utc + 1000.months) }
```
Maybe with modification above you will need to change this.

Drocheg committed:
Changes code style and fixes in book controller test

- Adds Faker for passwords generation in user factory
- Changes some single line do/end to single line {}
- Changes some let! to let

Drocheg committed:
Fix rubocop violations

Gmarn commented on spec/controllers/books_controller_spec.rb:
```diff
@@ -0,0 +1,111 @@
+      let(:book) { create(:book) }
+
+      let!(:rent) do
+        create(:rent, book: book, user: user, start_date: Time.zone.now - 20.months \
```
Please avoid this syntaxis, better something like that:
```ruby
create(:rent, book: book, 
              user: user, 
              start_date: Time.zone.now - 10.months,
              end_date: Time.zone.now + 10.months)
```

Gmarn replied:
Moreover her you do not need to call `start date` and `end date` because by default they will be in your factory

Gmarn commented on spec/controllers/books_controller_spec.rb:
```diff
@@ -0,0 +1,111 @@
+      let(:book) { create(:book) }
+
+      let!(:rent) do
+        create(:rent, book: book, user: user, start_date: Time.zone.now - 10.months \
```
Same

Gmarn commented on spec/factories.rb:
```diff
@@ -4,9 +4,24 @@
+  factory :rent do
+    user { create(:user) }
+    book { create(:book) }
+    start_date { Faker::Time.between(Time.now.utc - 1000.months, Time.now.utc + 1000.months) }
```
You can also use `Time.zone.now` instead of `Time.now.utc`

Drocheg committed:
Subdivide factories in multiple files and code style changes
